### ISSUE TITLE
Demo camera follow: Memory leak

### ISSUE THREAD

**Issue by westnordost** [sargunv/maplibre-compose#42]
I noticed that the performance in the follow-camera demo rapidly degrades (on Android)

https://github.com/user-attachments/assets/d53c4bfc-5874-4f22-83fa-580a3e8803d9

Looking at the log, one can see that every 8 seconds or so, the garbage collector frees about 4.7 MB of data, i.e. that's about 500 kB of data per second being created/thrown away (in Java).

<details>
<summary>Click to expand log</summary>
<pre>
2024-11-27 18:37:40.350  4021-4021  maplibre-compose        dev.sargunv.maplibrecompose.demoapp  I  Setting style URL
2024-11-27 18:37:40.452  4021-4021  maplibre-compose        dev.sargunv.maplibrecompose.demoapp  I  Style finished loading
2024-11-27 18:37:40.516  4021-4021  maplibre-compose        dev.sargunv.maplibrecompose.demoapp  I  Queuing source target for addition
2024-11-27 18:37:40.571  4021-4021  maplibre-compose        dev.sargunv.maplibrecompose.demoapp  I  Queuing layer target-shadow for addition at anchor Top, index 0
2024-11-27 18:37:40.670  4021-4021  maplibre-compose        dev.sargunv.maplibrecompose.demoapp  I  Queuing layer target-circle for addition at anchor Top, index 1
2024-11-27 18:37:40.671  4021-4021  maplibre-compose        dev.sargunv.maplibrecompose.demoapp  I  Adding source target
2024-11-27 18:37:40.671  4021-4021  maplibre-compose        dev.sargunv.maplibrecompose.demoapp  I  Initializing anchor Top with layer target-circle
2024-11-27 18:37:40.671  4021-4021  maplibre-compose        dev.sargunv.maplibrecompose.demoapp  I  Adding layer target-shadow below target-circle
2024-11-27 18:37:40.727  4021-4040  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=43.12ms min=2.32ms max=239.35ms count=19
2024-11-27 18:37:41.191  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=50.42ms min=11.99ms max=194.60ms count=17
2024-11-27 18:37:42.200  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=44.26ms min=24.72ms max=206.51ms count=22
2024-11-27 18:37:42.302  4021-4040  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=1574.91ms min=1574.91ms max=1574.91ms count=1
2024-11-27 18:37:43.250  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=149.06ms min=24.40ms max=573.46ms count=7
2024-11-27 18:37:43.333  4021-4040  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=12.67ms min=1.80ms max=42.42ms count=51
2024-11-27 18:37:44.274  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=255.86ms min=89.12ms max=715.88ms count=4
2024-11-27 18:37:44.353  4021-4040  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=6.19ms min=1.67ms max=46.14ms count=57
2024-11-27 18:37:44.904  4021-4027  compose.demoapp         dev.sargunv.maplibrecompose.demoapp  I  Background concurrent mark compact GC freed 3944KB AllocSpace bytes, 4(184KB) LOS objects, 49% free, 4834KB/9668KB, paused 1.124ms,16.142ms total 63.065ms
2024-11-27 18:37:45.302  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=73.33ms min=31.01ms max=169.93ms count=14
2024-11-27 18:37:45.364  4021-4040  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=13.38ms min=1.86ms max=34.75ms count=57
2024-11-27 18:37:46.324  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=49.80ms min=14.71ms max=388.50ms count=19
2024-11-27 18:37:47.346  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=23.92ms min=11.34ms max=71.08ms count=40
2024-11-27 18:37:48.346  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=62.22ms min=27.45ms max=382.91ms count=16
2024-11-27 18:37:49.575  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=116.06ms min=17.22ms max=433.10ms count=10
2024-11-27 18:37:50.579  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=108.85ms min=40.85ms max=455.51ms count=9
2024-11-27 18:37:51.830  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=197.79ms min=27.56ms max=526.49ms count=6
2024-11-27 18:37:52.598  4021-4027  compose.demoapp         dev.sargunv.maplibrecompose.demoapp  I  Background concurrent mark compact GC freed 4808KB AllocSpace bytes, 0(0B) LOS objects, 49% free, 4794KB/9588KB, paused 632us,7.700ms total 26.595ms
2024-11-27 18:37:52.912  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=82.87ms min=16.37ms max=574.37ms count=12
2024-11-27 18:37:54.347  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=178.06ms min=19.54ms max=628.63ms count=8
2024-11-27 18:37:56.053  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=168.56ms min=27.24ms max=711.12ms count=10
2024-11-27 18:37:57.080  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=117.84ms min=17.77ms max=719.73ms count=8
2024-11-27 18:37:58.802  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=243.40ms min=25.13ms max=777.16ms count=7
2024-11-27 18:37:59.874  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=204.34ms min=18.17ms max=855.36ms count=5
2024-11-27 18:38:00.265  4021-4027  compose.demoapp         dev.sargunv.maplibrecompose.demoapp  I  Background concurrent mark compact GC freed 4728KB AllocSpace bytes, 0(0B) LOS objects, 49% free, 4770KB/9540KB, paused 517us,6.367ms total 26.820ms
2024-11-27 18:38:00.956  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=211.44ms min=21.94ms max=926.05ms count=5
2024-11-27 18:38:02.031  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=205.25ms min=18.92ms max=878.46ms count=5
2024-11-27 18:38:03.095  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=207.33ms min=17.08ms max=908.89ms count=5
2024-11-27 18:38:04.169  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=209.12ms min=23.02ms max=905.44ms count=5
2024-11-27 18:38:05.257  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=211.10ms min=16.99ms max=933.64ms count=5
2024-11-27 18:38:06.389  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=224.85ms min=32.88ms max=973.74ms count=5
2024-11-27 18:38:07.544  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=224.34ms min=16.59ms max=999.85ms count=5
2024-11-27 18:38:07.946  4021-4027  compose.demoapp         dev.sargunv.maplibrecompose.demoapp  I  Background concurrent mark compact GC freed 4704KB AllocSpace bytes, 0(0B) LOS objects, 49% free, 4770KB/9540KB, paused 507us,5.283ms total 24.707ms
2024-11-27 18:38:08.829  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=244.17ms min=23.38ms max=1092.26ms count=5
2024-11-27 18:38:10.201  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=266.55ms min=22.92ms max=1195.76ms count=5
2024-11-27 18:38:11.433  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=243.43ms min=20.96ms max=1117.10ms count=5
2024-11-27 18:38:12.656  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=244.17ms min=23.14ms max=1114.52ms count=5
2024-11-27 18:38:13.944  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=255.47ms min=17.87ms max=1179.42ms count=5
2024-11-27 18:38:15.220  4021-4147  EGL_emulation           dev.sargunv.maplibrecompose.demoapp  D  app_time_stats: avg=251.38ms min=22.55ms max=1164.20ms count=5
2024-11-27 18:38:15.629  4021-4027  compose.demoapp         dev.sargunv.maplibrecompose.demoapp  I  Background concurrent mark compact GC freed 4708KB AllocSpace bytes, 0(0B) LOS objects, 49% free, 4766KB/9532KB, paused 486us,5.270ms total 24.017ms
</pre>
</details>

I let run the profiler a little while and saw that the Java memory usage was increasing constantly. As for *what* is being allocated, not sure if that is helpful, but most allocations were of ...

![grafik](https://github.com/user-attachments/assets/242f867a-e62c-439c-972f-f3a2b3f37779)

Trying the profiler with the edge-to-edge demo, the Java memory remains constant *as long as* one doesn't move the map. If one moves the map (e.g. just in a tiny circle), memory usage is increasing the same way, however, here, I see that about 90% of all allocations are `float[]`s.

Trying the same in my app (also using MapLibre-native on Android, but without Compose), I can reproduce the increasing memory usage on moving the map in just a tiny circle, (however, definitely more slowly than for the camera follow demo). Here, however, not 90% of all allocations are `float[]` but a mix of `float[]`, `String`, `Object[]`, `byte[]` and `int[]`.

---

So, in conclusion, there seems to be a (slight?) memory leak in MapLibre-native, but in maplibre-compose too, which adds up.

**Comment by sargunv** [comment]
this is helpful, thank you!

Curious, what kind of device are you running this on? I haven't seen such degraded performance on my end, but haven't tested with older devices so I likely have blind spots here.

I know there's some allocations happening each frame, but am unsure how much of the issue is that known stuff:
* when the camera moves, we read camera position and instantiate a `CameraPosition` to update the state
* when that demo updates the camera position
  * we do a `CameraPosition.copy()` for the new position to set
  * and then the `position` setter itself needs to instantiate a camera update to pass to maplibre-native-android

I'd be curious to know
* are those known allocations above the whole problem? I'd expect these short lived objects never make it out of the youngest generation of the garbage collector, but I'm not super familiar there
* can `onCameraMove` trigger multiple times per frame? maybe we're updating the state too often
* what else might be the problem, if those aren't it

**Comment by sargunv** [comment]
17000 expression allocations are interesting. Are expressions being reconstructed every recomposition, which itself is every frame? If so, we may need to rethink the expression API (immediate thought is, one could put them in factory functions)

**Comment by sargunv** [comment]
`float[]` allocations could also be the position vector converter in the demo app (used to animate the point we're following in the demo)

**Comment by sargunv** [comment]
> Trying the profiler with the edge-to-edge demo, the Java memory remains constant as long as one doesn't move the map. If one moves the map (e.g. just in a tiny circle), memory usage is increasing the same way, however, here, I see that about 90% of all allocations are float[]s.

maybe camera padding? unsure if we allocate a new padding array or just reuse the existing reference when a new CameraPosition is constructed

**Comment by westnordost** [comment]
> Curious, what kind of device are you running this on? 

Android API 35 x64 emulator on Windows 10. Don't worry too much about the degraded performance, it might be an EGL emulation issue. (But the memory issue is independent of EGL emulation of course.)

> Are expressions being reconstructed every recomposition, which itself is every frame?

I don't see why not. Remember, we talked about this in the chat: Each time a `*Layer` composable is called, for all the default parameters, new expression objects are created for these default parameters. So, we have ~50 new instantiations of `Expression`s for every call to e.g. `SymbolLayer`. All composable function in that block are called (i.e. all layer composable functions) when any data has changed.

So, regarding expressions, I can think of the following solutions:
- Constants for default values
- have `null` as default values
- put the whole style definition of a `*Layer` into a data class, have a default constant for that data class
```kotlin
public fun LineLayer(
  //...
  style = LineStyle.Default
)

data class LineStyle(
  color: Expression<Color> = Default.color,
  // ... all the other style parameters
) {
  companion object {
    val Default = IconStyle(
      color = const(Color.Black),
      //... all the other defaults
    )
  }
}
```

**Comment by sargunv** [comment]
I'm curious about whether the skippability of compose functions mitigates this. If I were to encapsulate my layer in a composable, and the animation login in a parent composable, would the compose runtime skip all my expression instantiations?

what's changing every frame is camera position State and source data. The source is an input into the layer. But the data set on the source isn't State, the reference to the source isn't changing, and so [strong skipping](https://developer.android.com/develop/ui/compose/performance/stability/strongskipping) _should_ consider that as nothing changing for the layer input, right?

prob worth writing some [benchmarks](https://developer.android.com/topic/performance/benchmarking/benchmarking-overview) around such use cases, see #29

**Comment by sargunv** [comment]
> Android API 35 x64 emulator on Windows 10.

Ah, my dev machine is ARM (apple silicon) so I'm entirely unfamiliar with x64 android emulation

**Comment by sargunv** [comment]
> as they are not directly visible (😕) from the function signature.

at least with default constants, one can click through to their definition (could even get creative and name the constant with its value for visibility)

**Comment by westnordost** [comment]
> and so [strong skipping](https://developer.android.com/develop/ui/compose/performance/stability/strongskipping) should consider that as nothing changing for the wrapper, right?

I didn't quite follow, but according to the docs, strong skipping is enabled by default on the newest Kotlin compiler version. However, for some reason, in my project, it seemed to be still off (despite me using the latest version of Kotlin compiler), I had to enable it manually. Maybe it was just some temporary gradle cache issue, or maybe the documentation is not quite right.

**Comment by westnordost** [comment]
> So, regarding expressions, I can think of the following solutions:

Heh, I have one more idea. The `const` function could do more than meets the eye. Think how Java manages strings. We could do something similar:

There could be a global internal cache of objects (`Map<Any?, Expression<*>>`) and the `const` function checks if the parameter is already in that map. If yes, it just returns the associated value. If not, it creates the Expression, adds it to the map and then returns it.

Lines of code+docs -wise, this would probably even be the shortest solution, but maybe its a bit too 🎩✨🐇? In any case, thread-safety on access to this map likely needs to be ensured.

**Comment by sargunv** [comment]
> However, for some reason, in my project, it seemed to be still off (despite me using the latest version of Kotlin compiler), I had to enable it manually.

interesting, how did you check that it was off? I've just assumed it's on but don't know how best to verify (other than just explicitly turning it on in the build script)

> I didn't quite follow

Right now we have this:

```kt
@Composable
fun CameraFollowDemo() = Column {
  // ... animation logic here
  MaplibreMap(modifier = Modifier.weight(1f), styleUrl = DEFAULT_STYLE, cameraState = cameraState) {
    val targetSource = rememberGeoJsonSource("target", Point(animatedPosition))

    CircleLayer(
      id = "target-shadow",
      source = targetSource,
      radius = const(13.0),
      color = const(Color.Black),
      blur = const(1.0),
      translate = point(0.0, 1.0),
    )

    CircleLayer(
      id = "target-circle",
      source = targetSource,
      radius = const(7.0),
      color = const(MaterialTheme.colorScheme.primary),
      strokeColor = const(Color.White),
      strokeWidth = const(3.0),
    )
  }
  // ... map controls here
}
```

The content of the map is recomposing every frame because it's reading `animatedPosition`, which is State. And on each recompose, `CircleLayer` is getting called with all its properties being allocated. But what if we encapsulate the layers:

```
@Composable
fun CameraFollowDemo() = Column {
  // ... animation logic here
  MaplibreMap(modifier = Modifier.weight(1f), styleUrl = DEFAULT_STYLE, cameraState = cameraState) {
    val targetSource = rememberGeoJsonSource("target", Point(animatedPosition))
    LocationPuck(targetSource)
  }
  // ... map controls here
}

@Composable
fun LocationPuck(locationSource: Source) {
  CircleLayer(
    id = "target-shadow",
    source = locationSource,
    radius = const(13.0),
    color = const(Color.Black),
    blur = const(1.0),
    translate = point(0.0, 1.0),
  )

  CircleLayer(
    id = "target-circle",
    source = locationSource,
    radius = const(7.0),
    color = const(MaterialTheme.colorScheme.primary),
    strokeColor = const(Color.White),
    strokeWidth = const(3.0),
  )
}
```

If I understand how Compose works, LocationPuck should be skippable and therefore shouldn't recompose every frame even though its parent does, avoiding all those unnecessary allocations

**Comment by sargunv** [comment]
Arrange by call stack is an interesting view. I measured the follow demo for ~27 seconds.

![CleanShot 2024-11-28 at 02 58 23@2x](https://github.com/user-attachments/assets/cf205472-3ce2-42f6-81cc-bb55406ec0dd)

Starting with ~233k allocations, I drilled down through the largest calls. Sure enough, 90% of it is in `performFrameDispatch`. Drilling deeper, about 60% of that (125k) in `performRecompose`. And here it's interesting:

![CleanShot 2024-11-28 at 03 01 01@2x](https://github.com/user-attachments/assets/12e7a4f7-62e3-4bab-8454-da23b374a5ee)

Sure the expression functions are a few thousand allocations each, but why is `rememberJsonSource` at 10k? It's because we're serializing the point (StringBuilder in there) every frame to pass to the source. Expressions make up another 25k plus 5k to box some Doubles, so ~30k from expressions.

Up above, CameraPosition isn't too bad (I'd expect it to be worse, since we're making one every frame right?). The animation has a bunch here, plus boxing some integers and some layout stuff.

![CleanShot 2024-11-28 at 03 06 44@2x](https://github.com/user-attachments/assets/14e1a228-2ff3-4df1-9a6c-0dc6c216ec98)

Up higher, the side effect (set camera position) has a bunch too, mostly from MapLibre itself, though a few thousand from the conversion (`toMLNCameraPosition`)

![CleanShot 2024-11-28 at 03 11 12@2x](https://github.com/user-attachments/assets/97dd1e72-3a90-435f-b242-1f08c1229ae8)

**Comment by sargunv** [comment]
tldr; recomposing every frame when our composables have complex data results in a ton of allocations

but the app itself doesn't drop frames when I'm not running it in profiling mode (and I expect release to be more performance). So I don't know whether the problem is really that bad? Definitely a ton of room for optimization here. Tomorrow I'm gonna do some experimentation with skipping

**Comment by sargunv** [comment]
> Heh, I have one more idea. The const function could do more than meets the eye. Think how Java manages strings. We could do something similar

An alternative/simpler version, just predefine some common values (0, 1, etc) and check against those before allocating a new Expression. This might work better if combined with overloading const for the different Number types, since I think Number results in boxing. Integer types can even pull from an array of predefined expressions for 0..n. Enum types can have all of theirs predefined too. All of this without the overhead of hashing and the threading concerns of caching. 

Still need to deal with geojson somehow (is translating to MLN objects better than stringifying? are there other source types we can provide that are optimized for commonly animated use cases? Though I expect updating a source every frame is not a super common use case outside of this demo

**Comment by westnordost** [comment]
>> However, for some reason, in my project, it seemed to be still off (despite me using the latest version of Kotlin compiler), I had to enable it manually.

> interesting, how did you check that it was off? 

In Android Studio, you have the [layout inspector](https://developer.android.com/develop/ui/compose/performance/stability/diagnose#layout-inspector), which amongst other things shows the recomposition count.

tl;dr: create a situation that would recompose on every frame when strong skipping is off, turn it off manually. Confirm in the layout inspector that it indeed recomposes every frame, then remove the strong skipping = off setting, check the layout inspector again.

---

In my app, I had some widget that would always recompose when anything else in the screen changed. E.g. there was an animation (on some other widget) and during that animation, the first widget would always recompose.

I learned this was due to the first widget having a lambda as a callback, and since the lambda is created in the compose function itself, on every call of the compose function, that lambda has a different identity (i.e. is a different object) and thus counts as "data changed, need to recompose". ([See stackoverflow question](https://stackoverflow.com/questions/69095595/why-recomposition-happens-when-call-viewmodel-in-a-callback)).

Anyway,strong skipping solves this without the developer needing to employ ugly workarounds like suggested in that stackoverflow question, so I upgraded the Kotlin compiler to 2.0.20, compiled and looked at the layout inspector again. No change! Then I activated strong skipping manually and voilá, that widget wasn't recomposed on every frame anymore.

**Comment by westnordost** [comment]
>> I didn't quite follow

> [...]
> If I understand how Compose works, LocationPuck should be skippable and therefore shouldn't recompose every frame even though its parent does, avoiding all those unnecessary allocations

I'd also think so.

**Comment by westnordost** [comment]
> is translating to MLN objects better than stringifying?

Internally in MapLibre, the GeoJson features provided from Java via JNI to native are actually converted to the native representation feature-by-feature.
I half-expected that the GeoJson features from Java would be converted to string and then the native part would again parse the GeoJson string, but that's not the case.

So, using the objects saves on saving to and parsing from a string. (On the other hand: more allocations due to the intermediate GeoJson data structure, so who knows 🤷‍♂️

**Comment by sargunv** [comment]
> Why, locationSource changes, hence that function is called.

The reference doesn't change. We call a setter on it to set the data, but it's not State. Compose doesn't know the data inside the source changed. Since strong skipping uses referential equality, I think compose should skip that function. I wish there was tooling to monitor non-UI compositions like we can for UI compositions

**Comment by sargunv** [comment]
> Depends on the device, though? We developers tend to have relatively new and high-end devices, but actually it would make sense to at least also test on old low-end devices, too.

Absolutely agree we should make an effort to support low end devices. **As an app dev**, I don't really want to optimize my apps beyond maybe ~4 year old mid range and ~2 year old low end devices. **But as a library dev**, I think it makes sense to optimize beyond that as far as the maintenance burden is reasonable, as there could be apps built on the library with target audiences using older devices. And on a mobile device, even if performance is good, reducing resource consumption reduces battery drain.

I'd just prefer to benchmark properly to know better the state of performance on such devices.

If we're profiling allocations, then lots of small allocations will have an outsized impact on performance due to the overhead of profiling. But the JVM is optimized* to handle use cases with large numbers of short-lived allocations (see generational garbage collection). So if we can make simple improvements to reduce allocations, that's worthwhile for sure. But if we're considering large refactors or complex mechanisms to avoid allocations, we should do that based on evidence from benchmarks with release-like builds, not profile-able ones.

*JVM is, but last I checked the GC for Kotlin Native does not split the heap into generations

**Comment by westnordost** [comment]
> As an app dev, I don't really want to optimize my apps beyond maybe ~4 year old low-mid range devices.

Off topic: I just got a bug [report](https://github.com/streetcomplete/StreetComplete/issues/6029) about the MapLibre map not rendering on his Samsung Galaxy S II 🤯 (That's almost 14 years old! 😵)

**Comment by sargunv** [comment]
Wow, thinking back to phones of that time (I started with an Evo 4G) they were janky even on software of their day. Can't imagine what's it's like running today's apps on them. Maybe I'll fire up my old Evo 4G or Galaxy Note 3 to see what it's like 😅

**Comment by boldtrn** [comment]
> Off topic: I just got a bug report about the MapLibre map not rendering on his Samsung Galaxy S II (released almost 14 years ago) 🤯

If you add the OpenGL 3.0 requirement to the Manifest, then they shouldn't be able to install your app through Google Play

**Comment by westnordost** [comment]
> If you add the OpenGL 3.0 requirement to the Manifest, then they shouldn't be able to install your app through Google Play

I already have. Looks like that device supports that. (Not sure if it is a hardware thing, the user put some newer LineageOS on that, of course)

**Comment by sargunv** [comment]
Merged some low hanging fruit to reduce allocations when Layer composables recompose. We're now reusing `const()` expressions for common discrete values like true, false, black, white, transparent, null, and small ints, so the vast majority of default values in Layer composables will no longer churn through the heap.

We can take that a step further with an LRU cache like what @westnordost described above (though we'd still allocate on boxing the key for the cache, we'd save the Expression allocation). One blocker there is that Kotlin's LinkedHashMap doesn't have the lru mode that the JVM version has, so we'd have to built or find an LRU implementation for Multiplatform. Short term, we could create some cache with a max size that just stops caching new keys after that size. We'd probably handle reasonable use cases with just that and no LRU eviction policy.

Looked into the camera; looks like we get the callback to set it no more than once a frame, so there's no straightforward optimization of the camera model adapter without a refactor of how camera state works. 

Same goes for passing data to GeoJsonSource every frame. Before refactoring these two things, I'd want to get some benchmarks and perf stories from real low end devices.

**Comment by sargunv** [comment]
More fun: I tested the theory of encapsulating the map layers into a separate composable, and looks like it does indeed prevent recompositions of the layers, preventing allocations of those expressions entirely (even without the above optimizations).

**Comment by sargunv** [comment]
With all the recent expression changes, 100% of the default layer property expressions should avoid allocation on recomposition. We're reusing constant instances for:
* all enums, booleans, nulls
* all numbers/dp that are multiples of 1f in 0..512
* all numbers/dp that are multiples of 0.05f in 0..25.6
* transparent, black, and white colors
* Offset.Zero and DpOffset.Zero
* some extras: default heatmap color interpolator, default font name list, default text padding (zero) 

Users can ofc define expressions beyond what we pre-generate, but users can also encapsulate their layers in skippable composables if that's a problem for them (or, `remember` complex expressions). So I'm gonna call this good for expression stuff.

Will create separate issues for investigating allocation impact and optimizations on camera movement and geojson data update.
